src/main/java/au/org/ala/biocache/dao/SearchDAOImpl.java

```diff
@@ -911,120 +1137,186 @@ public void writeCoordinatesToStream(SearchRequestParams searchParams,OutputStre
                             }
                         }
                     }
-                    allComplete = completeFutures.size() == futures.size();
-                    if(!allComplete){
-                        Thread.sleep(1000);
+                    if ((System.currentTimeMillis() - start) > downloadMaxTime) {
+                        interruptFound.set(true);
+                        break;
+                    }
+                    Thread.sleep(downloadCheckBusyWaitSleep);
+                } while (waitAgain);
+                
+                AtomicInteger totalDownload = new AtomicInteger(0);
+                for(Future<Integer> future: futures){
+                    if (future.isDone()){
+                        totalDownload.addAndGet(future.get());
+                    } else {
+                        // All incomplete futures that survived the loop above are cancelled here
+                        future.cancel(true);
                     }
                 }
-                pool.shutdown();
-                out.flush();
 
                 long finish = System.currentTimeMillis();
                 long timeTakenInSecs = (finish-start)/1000;
                 if(timeTakenInSecs <= 0) timeTakenInSecs = 1;
-                logger.info("Download of " + resultsCount + " records in " + timeTakenInSecs + " seconds. Record/sec: " + resultsCount.intValue()/timeTakenInSecs);
+                if(logger.isInfoEnabled()) {
+                    logger.info("Download of " + resultsCount + " records in " + timeTakenInSecs + " seconds. Record/sec: " + resultsCount.intValue()/timeTakenInSecs);
+                }
             } finally {
-                rw.finalise();
+                try {
+                    // Once we get here, we need to finalise starting at the concurrent wrapper, 
+                    // as there are no more non-sentinel records to be added to the queue
+                    // This eventually triggers finalisation of the underlying writer when the queue empties
+                    // This is a soft shutdown, and hence we wait below for this stage to complete in normal circumstances
+                    // Note, this blocks for writerTimeoutWaitMillis trying to legitimately add the sentinel to the end of the queue
+                    // We force the sentinel to be added in the hard shutdown phase below
+                    concurrentWrapper.finalise();
+                } finally {
+                    try {
+                        final long completionStartTime = System.currentTimeMillis();
+                        // Busy wait check for finalised to be called in the RecordWriter or something is interrupted
+                        // By this stage, there are at maximum download.internal.queue.size items remaining (default 1000)
+                        while(writerThread.isAlive() 
+                               && !writerThread.isInterrupted()
+                               && !interruptFound.get() 
```

Comment:
DownloadCreatorImpl has a catch(Exception) block that sends an email indicating failure, which can be triggered if Future.get throws an exception. 

However, as you say, there is a specific catch and log clause for SolrServerException in writeResultsFromIndexToStream and in some other methods. That can't be helped without a separate effort to remove the catch/log pattern with catch/log/rethrow that propagates the Exception instead of swallowing it after logging because it isn't just this one case that it occurs in. Probably best to get this pull request through before doing that task.